vmm/src/landlock.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright © 2024 Microsoft Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use landlock::{
+    path_beneath_rules, Access, AccessFs, BitFlags, PathFdError, Ruleset, RulesetAttr,
+    RulesetCreated, RulesetCreatedAttr, RulesetError, ABI,
+};
+use std::io::Error as IoError;
+use std::path::PathBuf;
+use thiserror::Error;
+
+use crate::vm_config::LandlockConfig;
+
+pub const READ: u8 = 1 << 0;
+pub const WRITE: u8 = 1 << 1;
+pub const EXECUTE: u8 = 1 << 2;
+
+#[derive(Debug, Error)]
+pub enum LandlockError {
+    /// Failed to create Ruleset
+    #[error("Error creating ruleset: {0}")]
+    CreateRuleSet(#[source] RulesetError),
+
+    /// Failed to add rule
+    #[error("Error adding rule to ruleset: {0}")]
+    AddRule(#[source] RulesetError),
+
+    /// Failed to restrict self
+    #[error("Error restricting self: {0}")]
+    RestrictSelf(#[source] RulesetError),
+
+    /// Ruleset is empty
+    #[error("Ruleset was previously consumed, cannot add rules to it")]
+    RulesetConsumed,
+
+    /// Error opening Path
+    #[error("Error opening Path: {0}")]
+    OpenPath(#[source] PathFdError),
+
+    /// Error opening Device Path
+    #[error("Error opening Device Path: {0}")]
+    OpenDevicePath(#[source] IoError),
+}
+
+pub struct Landlock {
+    ruleset: Option<RulesetCreated>,
+    abi: ABI,
+}
+
+impl Landlock {
+    pub fn new() -> Result<Landlock, LandlockError> {
+        let abi = ABI::V3;
+
+        let file_access = AccessFs::from_all(abi);
+        let def_ruleset = Ruleset::default()
+            .handle_access(file_access)
+            .map_err(LandlockError::CreateRuleSet)?;
+
+        let ruleset = def_ruleset.create().map_err(LandlockError::CreateRuleSet)?;
+
+        Ok(Landlock {
+            ruleset: Some(ruleset),
+            abi,
+        })
+    }
+
+    pub fn add_rule(
+        &mut self,
+        path: PathBuf,
+        access: BitFlags<AccessFs>,
+    ) -> Result<(), LandlockError> {
+        if self.ruleset.is_none() {
+            return Err(LandlockError::RulesetConsumed);
+        }
+        // path_beneath_rules in landlock crate handles file and directory access rules.
+        // Incoming path/s are passed to path_beneath_rules, so that we don't
+        // have to worry about the type of the path.
+        let paths = vec![path.clone()];
+        let path_beneath_rules = path_beneath_rules(paths, access);
+        for rule in path_beneath_rules {
+            self.ruleset
+                .as_mut()
+                .unwrap()
+                .add_rule(rule.unwrap())
+                .map_err(LandlockError::AddRule)?;
+        }
+        Ok(())
+    }
+
+    fn flags_to_access(&self, flags: u8) -> BitFlags<AccessFs> {
```

Comment:
Why using `u8` instead of an `enum` or a dedicated `struct` (following the builder pattern with a `read()` and `write()` method to change its internal private state)?

Reply:
`flags` here is an `OR` of of `READ`, `WRITE` and `EXECUTE`, which are declared as bits. 
After I read cmdline args like `--landlock-rules path=<a/b/c>,flags=rwx`, I translate the flags into transcode `rwx` to bitwise ORs of READ, WRITE, EXECUTE and pass it here.

Reply:
https://crates.io/crates/bitflags is a good shout for this

Reply:
Yes, but from an API point of view, the issue with this design is that unknown bits are ignored (at run time) instead of being denied by design (at build time).

Reply:
If unknown bits are passed at cmdline, cloud-hypervisor errors out in `LandlockConfig` -> `parse`.

Reply:
updated to use bitflags.